### ISSUE TITLE
Validation errors when using ordered_passes_renderpass!() to create a renderpass.

### ISSUE THREAD

**Issue by jonathansty** [vulkano-rs/vulkano#1063]
When trying to create a render pass using the `ordered_passes_renderpass` macro. I receive vulkan standard validation errors:
```
13:08:47 [ERROR] " [ VUID-VkSubpassDependency-srcAccessMask-00868 ] Object: VK_NULL_HANDLE (Type = 0) | CreateRenderPass: pDependencies[0].srcAccessMask (0x1FFFF) is not supported bysrcStageMask (0x8000). The Vulkan spec states: Any access flag included in srcAccessMask must be supported by one of the pipeline stages in srcStageMask, as specified in the table ofsupported access types (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-VkSubpassDependency-srcAccessMask-00868)"
13:08:47 [ERROR] " [ VUID-VkSubpassDependency-dstAccessMask-00869 ] Object: VK_NULL_HANDLE (Type = 0) | CreateRenderPass: pDependencies[0].dstAccessMask (0x1FFFF) is not supported bydstStageMask (0x8000). The Vulkan spec states: Any access flag included in dstAccessMask must be supported by one of the pipeline stages in dstStageMask, as specified in the table ofsupported access types (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-VkSubpassDependency-dstAccessMask-00869)"
13:08:47 [ERROR] " [ VUID-VkSubpassDependency-srcAccessMask-00868 ] Object: VK_NULL_HANDLE (Type = 0) | CreateRenderPass: pDependencies[1].srcAccessMask (0x1FFFF) is not supported bysrcStageMask (0x8000). The Vulkan spec states: Any access flag included in srcAccessMask must be supported by one of the pipeline stages in srcStageMask, as specified in the table ofsupported access types (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-VkSubpassDependency-srcAccessMask-00868)"
13:08:47 [ERROR] " [ VUID-VkSubpassDependency-dstAccessMask-00869 ] Object: VK_NULL_HANDLE (Type = 0) | CreateRenderPass: pDependencies[1].dstAccessMask (0x1FFFF) is not supported bydstStageMask (0x8000). The Vulkan spec states: Any access flag included in dstAccessMask must be supported by one of the pipeline stages in dstStageMask, as specified in the table ofsupported access types (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-VkSubpassDependency-dstAccessMask-00869)"
```

My code:
```
let render_pass = Arc::new(
    ordered_passes_renderpass!(
        gfx_queue.device().clone(),
        attachments: {
            // GBuffer attachments
            albedo : {
                load: Clear,
                store: DontCare,
                format: vulkano::format::Format::R8G8B8A8Unorm,
                samples: 1,
            },
            normal: {
                load: Clear,
                store: DontCare,
                format: vulkano::format::Format::R16G16B16A16Sfloat,
                samples: 1,
            },
            position: {
                load: Clear,
                store: DontCare,
                format: vulkano::format::Format::R16G16B16A16Sfloat,
                samples: 1,
            },

            // Composition output colors
            pp_color: {
                load: Clear,
                store: DontCare,
                format: swapchain.format(),
                samples: 1, 
            },
            final_color: {
                load: Clear,
                store: Store,
                format: swapchain.format(),
                samples: 1,
            },
            depth: {
                load: Clear,
                store: Store,
                format: vulkano::format::Format::D16Unorm,
                samples: 1,
            }
        },
        passes: [
            // GBuffer pass
            {
                color: [albedo,normal,position],
                depth_stencil: {depth},
                input: []
            },
            // Composition pass
            {
                color: [pp_color],
                depth_stencil: {},
                input: [albedo,normal,position]
            },
            // Post processing pass
            {
                color: [final_color],
                depth_stencil: {},
                input: [pp_color]
            }
        ]
    ).unwrap()
);
```

I've ran the deferred sample to check if it happens in a similar use case, mind you that I first had to enable any the standard validation layer and debug reporting for the instance. I can confirm that the deferred example also reports the same error messages. 

I think it is mainly due to the fact that the macro just sets all the source and access bits.
Snippet from `framebuffer/macros.rs`:
```
source_stages: PipelineStages { all_graphics: true, .. PipelineStages::none() },         // TODO: correct values
destination_stages: PipelineStages { all_graphics: true, .. PipelineStages::none() },         // TODO: correct values
source_access: AccessFlagBits::all(),         // TODO: correct values
destination_access: AccessFlagBits::all(),         // TODO: correct values
```

**Comment by jonathansty** [comment]
Seeing as I'm not super experienced using rust I decided to give a go to implementing my own render pass creation code which solves the validation errors completely. : 
```
    let swapchain_format = swapchain.format();

    struct DefaultRenderPassDesc{
        pub attachments : Vec<LayoutAttachmentDescription>,
        pub subpasses : Vec<LayoutPassDescription>,
        pub dependencies : Vec<LayoutPassDependencyDescription>,
    }

    use vulkano::framebuffer::RenderPassDesc;
    unsafe impl RenderPassDesc for DefaultRenderPassDesc {
            #[inline]
            fn num_attachments(&self) -> usize {
                self.attachments.len()
            }

            #[inline]
            fn attachment_desc(&self, id: usize) -> Option<LayoutAttachmentDescription> {
                if id < self.attachments.len(){
                    return Some(self.attachments[id].clone());
                }

                return None;
            }

            #[inline]
            fn num_subpasses(&self) -> usize {
                self.subpasses.len()
            }

            #[inline]
            fn subpass_desc(&self, id: usize) -> Option<LayoutPassDescription> {
                if id < self.subpasses.len() {
                    return Some(self.subpasses[id].clone());
                }
                return None;
            }

            #[inline]
            fn num_dependencies(&self) -> usize {
                self.dependencies.len()
            }

            #[inline]
            fn dependency_desc(&self, id: usize) -> Option<LayoutPassDependencyDescription> {
                if id < self.dependencies.len() {
                    return Some(self.dependencies[id].clone());
                }
                return None;
            }
    }

    unsafe impl RenderPassDescClearValues<Vec<ClearValue>> for DefaultRenderPassDesc {
        fn convert_clear_values(&self, values: Vec<ClearValue>) -> Box<Iterator<Item = ClearValue>> {
            // FIXME: safety checks
            Box::new(values.into_iter())
        }
    }

   let render_pass_description = {
       let mut attachments = Vec::new();

       // 0: Albedo 
       attachments.push(LayoutAttachmentDescription{
           format: Format::R8G8B8A8Unorm,
           samples: 1,
           load: LoadOp::Clear,
           store: StoreOp::DontCare,
           stencil_load: LoadOp::DontCare,
           stencil_store: StoreOp::DontCare,
           initial_layout: ImageLayout::Undefined,
           final_layout: ImageLayout::ColorAttachmentOptimal,
       });

       // 1: Normal 
       attachments.push(LayoutAttachmentDescription{
           format: Format::R16G16B16A16Sfloat,
           samples: 1,
           load: LoadOp::Clear,
           store: StoreOp::DontCare,
           stencil_load: LoadOp::DontCare,
           stencil_store: StoreOp::DontCare,
           initial_layout: ImageLayout::Undefined,
           final_layout: ImageLayout::ColorAttachmentOptimal,
       });

       // 2: Position 
       attachments.push(LayoutAttachmentDescription{
           format: Format::R16G16B16A16Sfloat,
           samples: 1,
           load: LoadOp::Clear,
           store: StoreOp::DontCare,
           stencil_load: LoadOp::DontCare,
           stencil_store: StoreOp::DontCare,
           initial_layout: ImageLayout::Undefined,
           final_layout: ImageLayout::ColorAttachmentOptimal,
       });

       // 3: PP Color 
       attachments.push(LayoutAttachmentDescription{
           format: Format::R8G8B8A8Unorm,
           samples: 1,
           load: LoadOp::Clear,
           store: StoreOp::DontCare,
           stencil_load: LoadOp::DontCare,
           stencil_store: StoreOp::DontCare,
           initial_layout: ImageLayout::Undefined,
           final_layout: ImageLayout::ColorAttachmentOptimal,
       });

       // 4: Final Color
       attachments.push(LayoutAttachmentDescription{
           format: swapchain_format,
           samples: 1,
           load: LoadOp::Clear,
           store: StoreOp::Store,
           stencil_load: LoadOp::Clear,
           stencil_store: StoreOp::Store,
           initial_layout: ImageLayout::Undefined,
           final_layout: ImageLayout::PresentSrc,
       });

       // 5: Depth
       attachments.push(LayoutAttachmentDescription{
           format: Format::D16Unorm,
           samples: 1,
           load: LoadOp::Clear,
           store: StoreOp::Store,
           stencil_load: LoadOp::Clear,
           stencil_store: StoreOp::Store,
           initial_layout: ImageLayout::Undefined,
           final_layout: ImageLayout::DepthStencilAttachmentOptimal,
       });

    let mut subpasses = Vec::new();
    // Deferred rendering to  GBuffer
    subpasses.push(LayoutPassDescription{
      color_attachments: vec![
          (0,ImageLayout::ColorAttachmentOptimal), // Albedo
          (1,ImageLayout::ColorAttachmentOptimal), // Normal
          (2,ImageLayout::ColorAttachmentOptimal), // Position
      ],
      depth_stencil: Some((5, ImageLayout::DepthStencilAttachmentOptimal)),
      input_attachments:     Vec::new(),
      resolve_attachments:   Vec::new(),
      preserve_attachments:  Vec::new()
    });

    // Composition 
    subpasses.push(LayoutPassDescription{
      color_attachments: vec![
          (3,ImageLayout::ColorAttachmentOptimal), 
      ],
      depth_stencil: None,
      input_attachments:     vec![
          (0,ImageLayout::ShaderReadOnlyOptimal), // Albedo
          (1,ImageLayout::ShaderReadOnlyOptimal), // Normal
          (2,ImageLayout::ShaderReadOnlyOptimal), // Position
      ],
      resolve_attachments:   Vec::new(),
      preserve_attachments:  Vec::new()
    });

    // Final post processing task
    subpasses.push(LayoutPassDescription{
      color_attachments: vec![
          (4,ImageLayout::ColorAttachmentOptimal), 
      ],
      depth_stencil: None,
      input_attachments:     vec![
          (3,ImageLayout::ShaderReadOnlyOptimal), // PP
      ],
      resolve_attachments:   Vec::new(),
      preserve_attachments:  Vec::new()
    });

    let mut dependencies = Vec::new();

    // ? -> deferred 
    dependencies.push(LayoutPassDependencyDescription{
        source_subpass:  vk::SUBPASS_EXTERNAL as usize,
        destination_subpass: 0,
        source_stages: PipelineStages{
            bottom_of_pipe: true,
            .. PipelineStages::none()
        },
        destination_stages: PipelineStages{
            color_attachment_output: true,
            .. PipelineStages::none()
        },
        source_access: AccessFlagBits{
            memory_read: true, 
            ..AccessFlagBits::none()
        },
        destination_access: AccessFlagBits{
            color_attachment_read:true, 
            color_attachment_write:true,
            ..AccessFlagBits::none()
        },
        by_region: true,
    });

    // Deferred -> composition
    dependencies.push(LayoutPassDependencyDescription{
        source_subpass: 0,
        destination_subpass: 1,
        source_stages: PipelineStages{
            color_attachment_output:true,
            ..PipelineStages::none()
        },
        destination_stages: PipelineStages{
            fragment_shader:true,
            .. PipelineStages::none()
        },
        source_access: AccessFlagBits{
            color_attachment_write:true,
            ..AccessFlagBits::none()
        },
        destination_access: AccessFlagBits{
            shader_read:true,
            ..AccessFlagBits::none()
        },
        by_region: true,
    });

    // Composition -> Final output (PP)
    dependencies.push(LayoutPassDependencyDescription{
        source_subpass: 1,
        destination_subpass: 2,
        source_stages: PipelineStages{
            color_attachment_output:true,
            ..PipelineStages::none()
        },
        destination_stages: PipelineStages{
            fragment_shader:true,
            .. PipelineStages::none()
        },
        source_access: AccessFlagBits{
            color_attachment_write:true,
            ..AccessFlagBits::none()
        },
        destination_access: AccessFlagBits{
            shader_read:true,
            ..AccessFlagBits::none()
        },
        by_region: true,
    });

    dependencies.push(LayoutPassDependencyDescription{
        source_subpass: 0,
        destination_subpass: vk::SUBPASS_EXTERNAL as usize,
        source_stages: PipelineStages{
            color_attachment_output:true,
            ..PipelineStages::none()
        },
        destination_stages: PipelineStages{
            bottom_of_pipe:true,
            .. PipelineStages::none()
        },
        source_access: AccessFlagBits{
            color_attachment_read:true,
            color_attachment_write:true,
            ..AccessFlagBits::none()
        },
        destination_access: AccessFlagBits{
            memory_read:true,
            ..AccessFlagBits::none()
        },
        by_region: true,
    });

    DefaultRenderPassDesc{
       attachments,
       subpasses,
       dependencies,
    }
   };

   vulkano::framebuffer::RenderPass::new(device, render_pass_description).unwrap()
```

**Comment by Rua** [comment]
This was fixed by #1828.
